hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java

```diff
@@ -852,49 +858,24 @@ private ClientConnection getConnectionFromLoadBalancer(boolean dataMember) {
         Address memberAddress = connection.getInitAddress();
         ClientMessage request = encodeAuthenticationRequest(memberAddress);
         ClientInvocationFuture future = new ClientInvocation(client, request, null, connection).invokeUrgent();
-        ClientAuthenticationCodec.ResponseParameters response;
         try {
-            response = ClientAuthenticationCodec.decodeResponse(future.get(authenticationTimeout, MILLISECONDS));
+            return ClientAuthenticationCodec.decodeResponse(future.get(authenticationTimeout, MILLISECONDS));
         } catch (Exception e) {
             connection.close("Failed to authenticate connection", e);
             throw rethrow(e);
         }
-
-        AuthenticationStatus authenticationStatus = AuthenticationStatus.getById(response.status);
-        if (failoverConfigProvided && !response.failoverSupported) {
-            logger.warning("Cluster does not support failover. This feature is available in Hazelcast Enterprise");
-            authenticationStatus = NOT_ALLOWED_IN_CLUSTER;
-        }
-        switch (authenticationStatus) {
-            case AUTHENTICATED:
-                return response;
-            case CREDENTIALS_FAILED:
-                AuthenticationException authException = new AuthenticationException("Authentication failed. The configured "
-                        + "cluster name on the client (see ClientConfig.setClusterName()) does not match the one configured in "
-                        + "the cluster or the credentials set in the Client security config could not be authenticated");
-                connection.close("Failed to authenticate connection", authException);
-                throw authException;
-            case NOT_ALLOWED_IN_CLUSTER:
-                ClientNotAllowedInClusterException notAllowedException =
-                        new ClientNotAllowedInClusterException("Client is not allowed in the cluster");
-                connection.close("Failed to authenticate connection", notAllowedException);
-                throw notAllowedException;
-            default:
-                AuthenticationException exception =
-                        new AuthenticationException("Authentication status code not supported. status: " + authenticationStatus);
-                connection.close("Failed to authenticate connection", exception);
-                throw exception;
-        }
     }
 
     /**
      * The returned connection could be different than the one passed to this method if there is already an existing
      * connection to the given member.
      */
     private TcpClientConnection onAuthenticated(TcpClientConnection connection,
-                                                ClientAuthenticationCodec.ResponseParameters response) {
+                                                ClientAuthenticationCodec.ResponseParameters response,
+                                                boolean switchingToNextCluster) {
```

Comment:
Can't you just use the new state `ClientState.SWITCHING_CLUSTER` and not have this new parameter `switchingToNextCluster `? Would it not be enough? 

Solution seems fine but in general, the logic is becoming more complex for the failover with a number of rules and it makes me to ask myself if there is a problem in our design.

Reply:
Unfortunately not. 
We had the `ClientState.SWITCHING_CLUSTER` state before. It was not enum but a boolean field, but information was there. 
That information itself was not enough to decide which connections should be authenticated or not. 
As I stated on the pr description, we need a second information to understand with which intent this connection is opened(Is it just opening regular connections via `ConnectToAllClusterMembersTask` or is failover logic kicked in).